File: src/ros2_bridge/mappers/sensor_msgs/multi_dof_joint_state.rs
//! Typed mapper for `sensor_msgs/msg/MultiDOFJointState`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn multi_dof_joint_state_to_bus(msg: ros_env::sensor_msgs::msg::MultiDOFJointState) -> crate::sensor_msgs::msg::v1::MultiDofJointState {
    crate::sensor_msgs::msg::v1::MultiDofJointState {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        joint_names: crate::ros2_bridge::mappers::convert::string_seq(msg.joint_names),
        transforms: msg.transforms.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::transform::transform_to_bus).collect(),
        twist: msg.twist.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::twist::twist_to_bus).collect(),
        wrench: msg.wrench.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::wrench::wrench_to_bus).collect(),
    }
}

pub(crate) fn multi_dof_joint_state_to_ros(bus: crate::sensor_msgs::msg::v1::MultiDofJointState) -> ros_env::sensor_msgs::msg::MultiDOFJointState {
    ros_env::sensor_msgs::msg::MultiDOFJointState {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        joint_names: crate::ros2_bridge::mappers::convert::ros_string_seq(bus.joint_names),
        transforms: bus.transforms.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::transform::transform_to_ros).collect(),
        twist: bus.twist.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::twist::twist_to_ros).collect(),
        wrench: bus.wrench.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::wrench::wrench_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsMultiDofJointStateMapper;

impl TypedTopicMapper for SensorMsgsMultiDofJointStateMapper {
    type Ros = ros_env::sensor_msgs::msg::MultiDOFJointState;
    type Bus = crate::sensor_msgs::msg::v1::MultiDofJointState;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(multi_dof_joint_state_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(multi_dof_joint_state_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/multi_dof_joint_state.py
"""Generated mapper for `sensor_msgs/msg/MultiDOFJointState`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.transform import transform_to_bus, transform_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.twist import twist_to_bus, twist_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.wrench import wrench_to_bus, wrench_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.sensor_msgs.msg.v1 import MultiDOFJointState as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def multi_dof_joint_state_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.joint_names.extend([str(x) for x in msg.joint_names])
    bus.transforms.extend([transform_to_bus(x) for x in msg.transforms])
    bus.twist.extend([twist_to_bus(x) for x in msg.twist])
    bus.wrench.extend([wrench_to_bus(x) for x in msg.wrench])
    return bus


def multi_dof_joint_state_to_ros(bus):
    from sensor_msgs.msg import MultiDOFJointState as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.joint_names = [str(x) for x in bus.joint_names]
    out.transforms = [transform_to_ros(x) for x in bus.transforms]
    out.twist = [twist_to_ros(x) for x in bus.twist]
    out.wrench = [wrench_to_ros(x) for x in bus.wrench]
    return out


class SensorMsgsMultiDofJointStateMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from sensor_msgs.msg import MultiDOFJointState as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return multi_dof_joint_state_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return multi_dof_joint_state_to_ros(bus)
